analyses/cms-open-data-ttbar/ml.py

```diff
@@ -0,0 +1,276 @@
+        .Define("Electron_eta_cut", "Electron_eta[Electron_mask]")
+        .Define(
+            "Electron_p4",
+            """
+            ConstructP4 (
+                    Electron_pt[Electron_mask], 
+                    Electron_eta_cut, 
+                    Electron_phi_cut, 
+                    Electron_mass[Electron_mask]
+            )
+            """,
+        )
+        .Define("Muon_phi_cut", "Muon_phi[Muon_mask]")
+        .Define("Muon_eta_cut", "Muon_eta[Muon_mask]")
+        .Define(
+            "Muon_p4",
+            """
+            ConstructP4 (
+                    Muon_pt[Muon_mask], 
+                    Muon_eta_cut, 
+                    Muon_phi_cut, 
+                    Muon_mass[Muon_mask]
+            )            
+            """,
+        )
+        .Define("Lepton_phi", "Concatenate(Electron_phi_cut, Muon_phi_cut)")
+        .Define("Lepton_eta", "Concatenate(Electron_eta_cut, Muon_eta_cut)")
+        .Define("Lepton_p4", "Concatenate(Electron_p4, Muon_p4)")
+    )
+
+    return df
+
+
+def define_features(df: ROOT.RDataFrame) -> ROOT.RDataFrame:
+    # prepare lepton fields (p4, eta, phi)
+    df = define_lepton_fields(df)
+
+    # get indexes of four jets
+    df = (
+        df.Define("W1_idx", "permutations[std::min(Jet_pt_cut.size(), max_n_jets)][0]")
+        .Define("W2_idx", "permutations[std::min(Jet_pt_cut.size(), max_n_jets)][1]")
+        .Define("bH_idx", "permutations[std::min(Jet_pt_cut.size(), max_n_jets)][2]")
+        .Define("bL_idx", "permutations[std::min(Jet_pt_cut.size(), max_n_jets)][3]")
+    )
+
+    feature_names = ml_feature_histo_config["names"]
+
+    # # Apply indexes to jets. Jets pt and btagCSVV2 and qgl are features itself (12 features)
+    df = (
+        df
+        # not features themself, but needed to construct features
+        .Define("JetW1_phi", "Take(Jet_phi_cut, W1_idx)")
+        .Define("JetW2_phi", "Take(Jet_phi_cut, W2_idx)")
+        .Define("JetbL_phi", "Take(Jet_phi_cut, bL_idx)")
+        .Define("JetbH_phi", "Take(Jet_phi_cut, bH_idx)")
+        .Define("JetW1_eta", "Take(Jet_eta_cut, W1_idx)")
+        .Define("JetW2_eta", "Take(Jet_eta_cut, W2_idx)")
+        .Define("JetbL_eta", "Take(Jet_eta_cut, bL_idx)")
+        .Define("JetbH_eta", "Take(Jet_eta_cut, bH_idx)")
+        #         # 12 features
+        .Define(f"{feature_names[8]}", "Take(Jet_pt_cut, W1_idx)")
+        .Define(f"{feature_names[9]}", "Take(Jet_pt_cut, W2_idx)")
+        .Define(f"{feature_names[10]}", "Take(Jet_pt_cut, bH_idx)")
+        .Define(f"{feature_names[11]}", "Take(Jet_pt_cut, bL_idx)")
+        .Define(f"{feature_names[12]}", "Take(Jet_btagCSVV2_cut, W1_idx)")
+        .Define(f"{feature_names[13]}", "Take(Jet_btagCSVV2_cut, W2_idx)")
+        .Define(f"{feature_names[14]}", "Take(Jet_btagCSVV2_cut, bH_idx)")
+        .Define(f"{feature_names[15]}", "Take(Jet_btagCSVV2_cut, bL_idx)")
+        .Define("Jet_qgl_cut", "Jet_qgl[Jet_mask]")
```

Comment:
can you lift this outside of the block that defines features with `feature_names[X]`?